### DISCUSSION TITLE
Security improvement: Replace stablelib with audited noble libraries

### DISCUSSION THREAD

**Discussion by paulmillr** [WalletConnect/walletconnect-monorepo#4035] [Ideas]
noble cryptography (https://paulmillr.com/noble/) is a high-security, auditable set of contained cryptographic libraries with following features:

- No dependencies, protection against supply chain attacks
- Auditable TypeScript / JS code
- Supported on all major platforms
- Releases are signed with PGP keys and built transparently with NPM provenance

Furthermore, noble-hashes were created with a grant by Ethereum Foundation. hashes and curves have been independently audited by 3rd party entity.

You have following deps:

https://github.com/WalletConnect/walletconnect-monorepo/blob/4ee518cf58b7aa0c59289999e7443918b9d14706/packages/utils/package.json#L33-L38

They can be replaced with:

```json
"@noble/ciphers": "0.1.3",
"@noble/curves": "1.1.0",
"@noble/hashes": "1.3.1",
```

The usage is as following:

```js
import { chacha20poly1305 } from "@noble/ciphers/chacha";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { x25519 } from "@noble/curves/ed25519";
import { randomBytes as random } from "@noble/hashes/utils";
```

**Comment by paulmillr** [comment]
I assume this repo is unmaintained and you don't really care, since there was no response for 2.5 months (#2272)


@pedrouid @ganchoradkov @0xAsimetriq @bkrem

**Comment by hazae41** [comment]
Sad to see no response to this

I will build a custom WalletConnect lib, it will include adapters for Noble, WebAssembly and Stablelib so you will be able to choose which one you prefer (and you will be able to do it at runtime, e.g. depending on the user connection speed, tradeoff performances against bundle size)

By the way, it will also support arbitrary transport protocols like WebSocket-over-Tor instead of only native WebSockets

EDIT: this might explain why this issue is ignored
<img width="500" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/4405263/54c427e3-9482-42a8-85af-b5cc9f75c15a" />

**Comment by pedrouid** [comment]
Hi @paulmillr @hazae41, can you guys chill out?

I have no connection to StableLib whatsoever 😅 so no need to start conspiracies just because I have one-line commit correcting a typo inside a comment: https://github.com/StableLib/stablelib/commit/cad7e89f9b4df831babc0342f17d1508c31a604e

Now it's important to note that my time in the last 18 months has not been spent coding or reviewing PRs so you are trying to reach out to the wrong guy

Also I follow both of you on Twitter and my DMs are open plus I've followed @paulmillr work since you started writing the Noble cryptography library

So I'm not committing to switch cryptography libraries in this thread but I promise you that I will ask the Javascript team to investigate the pros and cons

Give us some time to evaluate this and we can connect to discuss 

<img width="617" alt="Screenshot 2023-09-15 at 14 42 28" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/10136079/d43946d2-d657-45fe-8861-4cc0d96fe191">

Finally we take security very seriously but this comment above is not the right away to approach it and it just makes your argument weaker

**Comment by paulmillr** [comment]
Exactly — the issue is response time, which is 5 months at this point. The same reason makes angry security researchers drop 0days in public, without responsible disclosure.

**Comment by paulmillr** [comment]
@pedrouid 

>So I'm not committing to switch cryptography libraries in this thread but I promise you that I will ask the Javascript team to investigate the pros and cons

>Give us some time to evaluate this and we can connect to discuss

When can we expect a response from the team?

**Comment by paulmillr** [comment]
@pedrouid ?

**Comment by glitch-txs** [comment]
@paulmillr what's wrong with the current configuration? I'm aware noble is audited and well trusted.

**Reply by v1rtl** [reply]
stablelib is not audited by anyone nor used widely. if there's a vulnerability in stablelib it won't ever be noticed due to a tiny usage compared to noble hashes which are continiously reviewed and audited for security. https://github.com/paulmillr/noble-hashes/?tab=readme-ov-file#security

As for non-security reasons, using a yet-another-security-library bloats up dependency tree when everyone else is using noble-hashes/curves, thus needlessly increasing bundle size with duplicate implementations

hopefully those arguments are compelling enough

**Reply by hazae41** [reply]
incompetence or malice? who knows? ☹️

**Reply by hazae41** [reply]
there is some point where incompetence becomes malice when you blatantly ignore things

you guys should be ashamed of yourselves for not fixing such stuff on purpose and making the whole industry insecure

**Reply by arein** [reply]
@talentlessguy  thanks for raising this. When this library was audited by trail of bits this didn’t come up. We’ll ask some auditors and folks in the security community to inform what the right thing to do is and we’ll revert back here then 👍

**Reply by glitch-txs** [reply]
@talentlessguy we are more than happy to merge a PR on this. Otherwise we will look to address it in the near future. Thanks! 🙏🏻

**Reply by paulmillr** [reply]
> When this library was audited by trail of bits this didn’t come up

Because walletconnect did not ask trail of bits to audit dependencies.

[The audit pdf](https://github.com/trailofbits/publications/blob/master/reviews/2023-03-walletconnectv2-securityreview.pdf) is clear (page 9):

> we did not perform a dedicated code review of the dependencies themselves

**Comment by paulmillr** [comment]
Submitted a pull request: https://github.com/WalletConnect/walletconnect-monorepo/pull/5241

**Reply by paulmillr** [reply]
And second one https://github.com/WalletConnect/walletconnect-utils/pull/185

**Reply by glitch-txs** [reply]
Thank you 🙌🏻
